packages/nango-yaml/lib/parser.v2.ts

```diff
@@ -0,0 +1,261 @@
+import type {
+    HTTP_VERB,
+    NangoModel,
+    NangoSyncEndpoint,
+    NangoYamlParsedIntegration,
+    NangoYamlV2,
+    NangoYamlV2Integration,
+    ParsedNangoAction,
+    ParsedNangoSync
+} from '@nangohq/types';
+import { NangoYamlParser } from './parser.js';
+import {
+    ParserErrorDuplicateEndpoint,
+    ParserErrorDuplicateModel,
+    ParserErrorEndpointsMismatch,
+    ParserErrorInvalidRuns,
+    ParserErrorModelIsLiteral
+} from './errors.js';
+import { getInterval } from './helpers.js';
+
+export class NangoYamlParserV2 extends NangoYamlParser {
+    parse(): boolean {
+        const yaml = this.raw as unknown as NangoYamlV2;
+        const output: NangoYamlParsedIntegration[] = [];
+        this.modelsParser.parseAll();
+
+        if (this.modelsParser.errors.length > 0) {
+            this.errors.push(...this.modelsParser.errors);
+            return false;
+        }
+        if (this.modelsParser.warnings.length > 0) {
+            this.warnings.push(...this.modelsParser.warnings);
+        }
+
+        for (const integrationName in yaml.integrations) {
+            const integration = yaml.integrations[integrationName];
+            if (!integration) {
+                continue;
+            }
+
+            // check that models are used only once per integration
+            const usedModels = new Set<string>();
```

Comment:
I don't really understand the use of this mutable set `usedModels` passed around `parseSyncs` and `parseActions` can you not get the models from the parsed result and then check for errors (duplication, etc...)?

Reply:
resolved below